### PR TITLE
Mutual exclusivity for autologged runs

### PR DESCRIPTION
**Pull request by apurva-koti**
## What changes are proposed in this pull request?

Implement logic for allowing mutual exclusivity between autolog-created and user-created runs,
according to the user's specification in the `autolog(exclusive=False)` flag.

## How is this patch tested?

Unit tests

### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [x] Yes. Give a description of this change to be included in the release notes for MLflow users.

(Details in 1-2 sentences. You can just refer to another PR with a description if this PR is part of a larger change.)

### What component(s), interfaces, languages, and integrations does this PR affect?
Components 
- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: Local serving, model deployment tools, spark UDFs
- [ ] `area/server-infra`: MLflow server, JavaScript dev server
- [x] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface 
- [ ] `area/uiux`: Front-end, user experience, JavaScript, plotting
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language 
- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations
- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [x] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: mlflow/utils/autologging_utils.py
Comment (apurva-koti) [inline mlflow/utils/autologging_utils.py:670]: Comments to be removed

File: mlflow/utils/autologging_utils.py
Comment (dbczumar) [inline mlflow/utils/autologging_utils.py:481]:
```suggestion
    _session = None
```

**Review by dbczumar — COMMENTED** [review]
@apurva-koti this is awesome! seems to be coming along really well! Can we also add `exclusive=False` to `mlflow.autolog()` and the existing integrations that use `safe_patch` Currently: `mlflow.sklearn`, `mlflow.xgboost`, `mlflow.keras`. Note that we won't be adding this to `mlflow.spark`, since `exclusive` doesn't make sense for that use case.

File: mlflow/keras.py
Comment (dbczumar) [inline mlflow/keras.py:620]:
```suggestion
    :param exclusive: If ``True``, autologged content is not added to user-created fluent runs. If ``False``,
                                  autologged content is logged to the fluent active run, which may be user-created.
```

**Review by dbczumar — COMMENTED** [review]
@apurva-koti This is awesome! I tested it out manually and confirmed that `exclusive` works as expected. One final ask: Can we add an integration test to `tests/autologging/test_autologging_safety_integration.py` that picks a framework of your choosing (e.g., scikit-learn) and verifies that exclusive mode works properly for a simple example in that framework? Ideally, we should have done this for `disable=True/False` as well - I'll take care of that in a subsequent PR. After that, this LGTM!

File: mlflow/utils/autologging_utils.py
Comment (dbczumar) [inline mlflow/utils/autologging_utils.py:648]: Should we move this up next to the disablement check on lines 628 / 629?

File: tests/autologging/test_autologging_safety_unit.py
Comment (apurva-koti) [reply tests/autologging/test_autologging_safety_unit.py:948]: Makes sense. Let's do that

Comment (apurva-koti) [reply tests/autologging/test_autologging_safety_unit.py:948]: 💯

File: mlflow/utils/autologging_utils.py
Comment (apurva-koti) [reply mlflow/utils/autologging_utils.py:648]: Done!

**Review by dbczumar — APPROVED** [review]
LGTM once merge conflicts are resolved & tests pass! Awesome work @apurva-koti !

**Comment by dbczumar** [comment]
Failures in the windows & python small/large tests are attributable to the latest version of sqlalchemy (https://pypi.org/project/SQLAlchemy/) breaking our SQL schema consistency check. These failures are entirely independent of this PR. Accordingly, I'm going to merge this while we await a fix for the sql test failures.

**Pull request merged**
